navigator/runtime-experimental/src/main/java/com/freeletics/mad/navigator/compose/internal/MultiStack.kt

```diff
@@ -0,0 +1,181 @@
+    }
+
+    fun push(route: NavRoute) {
+        currentStack.push(route)
+        updateVisibleDestinations()
+    }
+
+    fun push(root: NavRoot, restoreRootState: Boolean, saveCurrentRootState: Boolean) {
+        val stack = getOrCreateBackStack(root)
+        switchToBackStack(stack, restoreRootState, saveCurrentRootState)
+        updateVisibleDestinations()
+    }
+
+    fun popCurrentStack() {
+        currentStack.pop()
+        updateVisibleDestinations()
+    }
+
+    fun pop() {
+        if (currentStack.isAtRoot) {
+            check(currentStack.id != startStack.id) {
+                "Can't navigate back from the root of the start back stack"
+            }
+            switchToBackStack(startStack, restoreRootState = false, saveCurrentRootState = false)
+        } else {
+            currentStack.pop()
+        }
+        updateVisibleDestinations()
+    }
+
+    fun <T : BaseRoute> popBackTo(
+        destinationId: DestinationId<T>,
+        isInclusive: Boolean
+    ) {
+        currentStack.popUpTo(destinationId, isInclusive)
+        updateVisibleDestinations()
+    }
+
+    fun saveState(): Bundle {
+        return bundleOf(
+            SAVED_STATE_ALL_STACKS to ArrayList(allStacks.map { it.saveState() }),
+            SAVED_STATE_CURRENT_STACK to currentStack.id.route.java,
+        )
+    }
+
+    companion object {
+        fun createWith(
+            root: NavRoot,
+            destinations: List<ContentDestination<*>>,
+            onStackEntryRemoved: (StackEntry.Id) -> Unit,
+            idGenerator: () -> String = { UUID.randomUUID().toString() },
+        ): MultiStack {
+            val startStack = Stack.createWith(root, destinations, onStackEntryRemoved, idGenerator)
+            return MultiStack(
+                allStacks = mutableListOf(startStack),
+                startStack = startStack,
+                currentStack = startStack,
+                destinations = destinations,
+                onStackEntryRemoved = onStackEntryRemoved,
+                idGenerator = idGenerator,
+            )
+        }
+
+        @Suppress("DEPRECATION")
+        fun fromState(
+            root: NavRoot,
+            bundle: Bundle,
+            destinations: List<ContentDestination<*>>,
+            onStackEntryRemoved: (StackEntry.Id) -> Unit,
+            idGenerator: () -> String = { UUID.randomUUID().toString() },
+        ): MultiStack {
+            val allStackBundles = bundle.getParcelableArrayList<Bundle>(SAVED_STATE_ALL_STACKS)!!
+            val currentStackId = bundle.getSerializable(SAVED_STATE_CURRENT_STACK)
+            val allStacks = allStackBundles.mapTo(ArrayList(allStackBundles.size)) {
+                Stack.fromState(it, destinations, onStackEntryRemoved, idGenerator)
+            }
+            val startStack = allStacks.find { it.id == root.destinationId }!!
+            val currentStack = allStacks.find { it.id.route.java == currentStackId }!!
```

Comment:
How about `first { ... }` is instead of `find  { ... }!!` 😁

Reply:
Yes, will update it 👍

Reply:
[c050a9c](https://github.com/freeletics/mad/pull/372/commits/c050a9cad75665333d284415eb31a2e86278f623)